python-lib/openweathermap_utils/utils.py

```diff
@@ -0,0 +1,184 @@
+    :return: New time floored
+    """
+    datetime_els = ['year', 'month', 'day', 'hour', 'minute', 'second', 'microsecond']
+    if not round_to in datetime_els:
+        raise KeyError(f'Error in flooring dt: You have to choose a value between these: {datetime_els}')
+    rt_index = datetime_els.index(round_to)
+    return dt.replace(**{k: 0 if rt_index + i > 1 else 1 for i, k in enumerate(datetime_els[rt_index + 1:])})
+
+
+def datetime_to_timestamp(dt):
+    return int(dt.timestamp())
+
+
+def timestamp_to_datetime(ts):
+    return datetime.fromtimestamp(int(ts))
+
+
+def datetime_to_str(dt, formatter="%Y-%m-%d"):
+    return dt.strftime(formatter)
+
+
+def str_to_datetime(date_str, formatter="%Y-%m-%d"):
+    return datetime.strptime(date_str, formatter)
+
+
+def make_column_names_unique(df):
+    cols = pd.Series(df.columns)
+    for dup in cols[cols.duplicated()].unique():
+        cols[cols[cols == dup].index.values.tolist()] = [
+            dup + '_' + str(i) if i != 0 else dup for i in range(sum(cols == dup))
+        ]
+    df.columns = cols
+
+
+def flatten_dict_rec(el, names, new_dict):
+    if not el:
+        new_dict['.'.join(names)] = ''
+    if isinstance(el, dict):
+        for k, v in el.items():
+            flatten_dict_rec(v, names + [str(k)], new_dict)
+    elif isinstance(el, list):
+        for i, e in enumerate(el):
+            flatten_dict_rec(e, names + [str(i)], new_dict)
+    else:
+        new_dict['.'.join(names)] = el
+
+
+def flatten_dict(d):
+    new_dict = {}
+    flatten_dict_rec(d, [], new_dict)
+    return new_dict
+
+
+def get_cache_location_from_configs(cache_location, default):
+    home_dir = pwd.getpwuid(os.getuid()).pw_dir
+    # Only solution to get user $HOME directory (getpass.getuser()
+    # doesn't work). We cannot write in the data_dir because of
+    # the implementation of MUS instances where permission problems
+    # could happen.
+    if cache_location == 'original':
+        return os.path.join(home_dir, constants.CACHE_RELATIVE_DIR)
+    elif cache_location == 'none':
+        return ''
+    return default
+
+
+def requests_error_handler(function):
+    """
+    Raises an exception or ignore it according to the type of error.
+    :param function: Function decorated
+    :return: (data, error) if error ignore else raises an Exception. If there is no error, the second argument is empty
+    """
+    def wrapper(*args, **kwargs):
+        try:
+            res = function(*args, **kwargs)
+            return res, OpenWeatherMapAPIError(status_code=200, text='')
+        except OpenWeatherMapAPIError as owm_error:
+            return [{}], owm_error
+        except HTTPError as http_error:
+            owm_error = OpenWeatherMapAPIError(
+                http_error.response.text,
+                status_code=int(http_error.response.status_code),
+                text=http_error.response.text
+            )
+            if owm_error.status_code == 401:
```

Comment:
What about 403?

Reply:
OpenWeatherMap raises a 401 error when the API key is not authorized to call a service. I don't know if it ever raises a 403 but added it in d3e0af3

Reply:
Ok, don't need to add a 403 so.